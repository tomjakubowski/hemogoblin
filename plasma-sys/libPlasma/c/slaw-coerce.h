
/* (c)  oblong industries */

#ifndef SLAW_COERCE_INTERROGATION
#define SLAW_COERCE_INTERROGATION

#include "libLoam/c/ob-types.h"
#include "libPlasma/c/slaw.h"

#ifdef __cplusplus
extern "C" {
#endif

/**                 Attempts to interpret the given slaw as an unt64 if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type, it will be converted to an unt64.
 *                  If s is a string, the string will be
 *                  parsed as an unt64 if possible.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets integer if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt64 (bslaw s, unt64 *result);

/**                 Attempts to interpret the given slaw as an unt32 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_unt64, which see, and the result is cast to
 *                  unt32.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt32 (bslaw s, unt32 *result);

/**                 Attempts to interpret the given slaw as an unt16 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_unt16, which see, and the result is cast to
 *                  unt16.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt16 (bslaw s, unt16 *result);

/**                 Attempts to interpret the given slaw as an unt8 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_unt8, which see, and the result is cast to
 *                  unt8.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_unt8 (bslaw s, unt8 *result);

/**                 Attempts to interpret the given slaw as an int64 if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type, it will be converted to an int64.
 *                  If s is a string, the string will be
 *                  parsed as an int64 if possible.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets integer if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int64 (bslaw s, int64 *result);

/**                 Attempts to interpret the given slaw as an int32 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int64, which see, and the result is cast to
 *                  int32.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int32 (bslaw s, int32 *result);

/**                 Attempts to interpret the given slaw as an int16 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int16, which see, and the result is cast to
 *                  int16.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int16 (bslaw s, int16 *result);

/**                 Attempts to interpret the given slaw as an int8 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int8, which see, and the result is cast to
 *                  int8.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_int8 (bslaw s, int8 *result);

/**                 Attempts to interpret the given slaw as a float64 if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type, it will be converted to a float64.
 *                  If s is a string, the string will be
 *                  parsed as a float64 if possible.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets float if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_float64 (bslaw s, float64 *result);

/**                 Attempts to interpret the given slaw as an float32 if
 *                  at all possible. The conversion is performed by @ref
 *                  slaw_to_int64, which see, and the result is cast to
 *                  float32.
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_float32 (bslaw s, float32 *result);

/**                 Attempts to interpret the given slaw as a length-2
 *                  vector if at all possible.  In particular, if s is a
 *                  list of the proper length, then the elements of the
 *                  list will be coerced as by slaw_to_float64.
 *                  If s is either an array or a vector of any real type,
 *                  of the proper length, it will be coerced to a vector
 *                  of float64.  And if s is a string, it will be parsed
 *                  as numbers separated by commas, if possible.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_v2 (bslaw s, v2float64 *result);

/**                 Attempts to interpret the given slaw as a length-3
 *                  vector if at all possible.  In particular, if s is a
 *                  list of the proper length, then the elements of the
 *                  list will be coerced as by slaw_to_float64.
 *                  If s is either an array or a vector of any real type,
 *                  of the proper length, it will be coerced to a vector
 *                  of float64.  And if s is a string, it will be parsed
 *                  as numbers separated by commas, if possible.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_v3 (bslaw s, v3float64 *result);

/**                 Attempts to interpret the given slaw as a length-4
 *                  vector if at all possible.  In particular, if s is a
 *                  list of the proper length, then the elements of the
 *                  list will be coerced as by slaw_to_float64.
 *                  If s is either an array or a vector of any real type,
 *                  of the proper length, it will be coerced to a vector
 *                  of float64.  And if s is a string, it will be parsed
 *                  as numbers separated by commas, if possible.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_v4 (bslaw s, v4float64 *result);

/**                 Implements the same coercion behavior as slaw_to_v2(),
 *                  slaw_to_v3(), and slaw_to_v4(), but for an arbitrary
 *                  length.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets the vector elements if return value is OB_OK,
 *                  otherwise unchanged
 *
 * \param[in]       capacity is the maximum number of float64s which can
 *                  be written to \a result.
 *
 * \param[out]      len is the number of vector elements actually written
 *                  to \a result, which will be less than or equal to
 *                  \a capacity.
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, SLAW_WRONG_LENGTH,
 *                  or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_vn (bslaw s, float64 *result, int capacity,
                                    int *len);


/**       Attempts to interpret the given slaw as a v ## N ## T, by
 *         first calling slaw_to_v ## N and then casting the
 *         components of the result to T
 */
#define DECLARE_V_COERCION(N, T)                                               \
  OB_PLASMA_API ob_retort slaw_to_v##N##T (bslaw s, v##N##T *result)

#define FOR_ALL_NUMERIC_TYPES(N, M)                                            \
  M (N, int8);                                                                 \
  M (N, int16);                                                                \
  M (N, int32);                                                                \
  M (N, int64);                                                                \
  M (N, unt8);                                                                 \
  M (N, unt16);                                                                \
  M (N, unt32);                                                                \
  M (N, unt64);                                                                \
  M (N, float32);                                                              \
  M (N, float64)

FOR_ALL_NUMERIC_TYPES (2, DECLARE_V_COERCION);
FOR_ALL_NUMERIC_TYPES (3, DECLARE_V_COERCION);
FOR_ALL_NUMERIC_TYPES (4, DECLARE_V_COERCION);

#undef FOR_ALL_NUMERIC_TYPES
#undef DECLARE_V_COERCION

/**                 Attempts to interpret the given slaw as a boolean if
 *                  at all possible.  In particular, if s is any atomic
 *                  numeric type with a value of "0" or "1", it will
 *                  be converted to false or true, respectively.
 *                  If s is a string, it may be either "true" or "false",
 *                  case-insensitively.
 *                  Is s is a list of one element or an array of one element,
 *                  then that one element shall be treated to the same
 *                  opportunities for coercion as described above.
 *
 * \param[in]       s is the slaw to be coerced
 *
 * \param[out]      result gets boolean if return value is OB_OK,
 *                  no guarantee otherwise
 *
 * \return          one of OB_OK, SLAW_RANGE_ERR, or SLAW_NOT_NUMERIC
 */
OB_PLASMA_API ob_retort slaw_to_boolean (bslaw s, bool *result);

#ifdef __cplusplus
}
#endif

#endif /* SLAW_COERCE_INTERROGATION */
