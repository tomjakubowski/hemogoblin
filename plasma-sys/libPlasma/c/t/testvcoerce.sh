#!/bin/bash
srcdir="$(cd "$(dirname "$0")" && pwd)"

# Handle internal valgrind
IV=""
if [ "$#" == "2" ]; then
    if [ "$1" == "--internal-valgrind" ]; then
        IV=$2
    fi
fi

PATH=.:$PATH

$IV \
testvcoerce > scratch/testvcoerce.out
[ "$?" != "0" ] && exit 1

exec diff -ub scratch/testvcoerce.out $srcdir/testvcoerce.expected
