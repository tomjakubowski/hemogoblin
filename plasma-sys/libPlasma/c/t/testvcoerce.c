
/* (c)  oblong industries */

#include "libLoam/c/ob-log.h"
#include "libPlasma/c/slaw-coerce.h"
#include "libPlasma/c/pool.h"
#include <stdio.h>
#include <stdlib.h>

static void trycoerce (int trial, slaw s)
{
  v2float64 v2;
  v3float64 v3;
  v4float64 v4;
  ob_retort err;

  printf ("\n===== test %d =====\n\n", trial);

  err = slaw_to_v2 (s, &v2);
  if (err == OB_OK)
    {
      printf ("slaw_to_v2: %f, %f\n", v2.x, v2.y);
    }
  else
    {
      printf ("slaw_to_v2: error %" OB_FMT_RETORT "d"
              "\n",
              err);
    }

  err = slaw_to_v3 (s, &v3);
  if (err == OB_OK)
    {
      printf ("slaw_to_v3: %f, %f, %f\n", v3.x, v3.y, v3.z);
    }
  else
    {
      printf ("slaw_to_v3: error %" OB_FMT_RETORT "d"
              "\n",
              err);
    }

  err = slaw_to_v4 (s, &v4);
  if (err == OB_OK)
    {
      printf ("slaw_to_v4: %f, %f, %f, %f\n", v4.x, v4.y, v4.z, v4.w);
    }
  else
    {
      printf ("slaw_to_v4: error %" OB_FMT_RETORT "d"
              "\n",
              err);
    }

  slaw_free (s);
}

int main (int argc, char **argv)
{
  v2int8 q1;
  v3float64 q2;
  v4float32 q3;
  v4unt64c q4;

  slabu *sb = slabu_new ();

  q1.x = 1;
  q1.y = 2;

  q2.x = 3;
  q2.y = 4;
  q2.z = 5;

  q3.x = 6;
  q3.y = 7;
  q3.z = 8;
  q3.w = 9;

  q4.x.re = 10;
  q4.x.im = 11;
  q4.y.re = 12;
  q4.y.im = 13;
  q4.z.re = 14;
  q4.z.im = 15;
  q4.w.re = 16;
  q4.w.im = 17;

  OB_DIE_ON_ERROR (slabu_list_add_x (sb, slaw_unt64 (1)));
  OB_DIE_ON_ERROR (slabu_list_add_x (sb, slaw_int8 (2)));
  OB_DIE_ON_ERROR (slabu_list_add_x (sb, slaw_float32 (3)));
  OB_DIE_ON_ERROR (slabu_list_add_x (sb, slaw_string ("4")));

  trycoerce (1, slaw_string ("1.2, -4.5, 7.8"));
  trycoerce (2, slaw_int32_array_filled (0, 36));
  trycoerce (3, slaw_int32_array_filled (1, 36));
  trycoerce (4, slaw_int32_array_filled (2, 36));
  trycoerce (5, slaw_int32_array_filled (3, 36));
  trycoerce (6, slaw_int32_array_filled (4, 36));
  trycoerce (7, slaw_int32_array_filled (5, 36));
  trycoerce (8, slaw_unt8_array_filled (0, 72));
  trycoerce (9, slaw_unt8_array_filled (1, 72));
  trycoerce (10, slaw_unt8_array_filled (2, 72));
  trycoerce (11, slaw_unt8_array_filled (3, 72));
  trycoerce (12, slaw_unt8_array_filled (4, 72));
  trycoerce (13, slaw_unt8_array_filled (5, 72));
  trycoerce (14, slaw_float32_array_filled (0, 3.14f));
  trycoerce (15, slaw_float32_array_filled (1, 3.14f));
  trycoerce (16, slaw_float32_array_filled (2, 3.14f));
  trycoerce (17, slaw_float32_array_filled (3, 3.14f));
  trycoerce (18, slaw_float32_array_filled (4, 3.14f));
  trycoerce (19, slaw_float32_array_filled (5, 3.14f));
  trycoerce (20, slaw_float64_array_filled (0, 3.14159));
  trycoerce (21, slaw_float64_array_filled (1, 3.14159));
  trycoerce (22, slaw_float64_array_filled (2, 3.14159));
  trycoerce (23, slaw_float64_array_filled (3, 3.14159));
  trycoerce (24, slaw_float64_array_filled (4, 3.14159));
  trycoerce (25, slaw_float64_array_filled (5, 3.14159));
  trycoerce (26, slaw_string ("8,9,10,11"));
  trycoerce (27, slaw_string ("8,9,10,11,12"));
  trycoerce (28, slaw_string ("-1, 0.0001"));
  trycoerce (29, slaw_nil ());
  trycoerce (30, slaw_string ("blech, foo, barf"));
  trycoerce (31, slaw_v2int8 (q1));
  trycoerce (32, slaw_v3float64 (q2));
  trycoerce (33, slaw_v4float32 (q3));
  trycoerce (34, slaw_v4unt64c (q4));
  trycoerce (35, slaw_list_f (sb));

  return EXIT_SUCCESS;
}
