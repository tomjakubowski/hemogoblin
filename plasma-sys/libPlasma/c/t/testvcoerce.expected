
===== test 1 =====

slaw_to_v2: error -210006
slaw_to_v3: 1.200000, -4.500000, 7.800000
slaw_to_v4: error -210006

===== test 2 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 3 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 4 =====

slaw_to_v2: 36.000000, 36.000000
slaw_to_v3: 36.000000, 36.000000, 0.000000
slaw_to_v4: error -210006

===== test 5 =====

slaw_to_v2: error -210006
slaw_to_v3: 36.000000, 36.000000, 36.000000
slaw_to_v4: error -210006

===== test 6 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: 36.000000, 36.000000, 36.000000, 36.000000

===== test 7 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 8 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 9 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 10 =====

slaw_to_v2: 72.000000, 72.000000
slaw_to_v3: 72.000000, 72.000000, 0.000000
slaw_to_v4: error -210006

===== test 11 =====

slaw_to_v2: error -210006
slaw_to_v3: 72.000000, 72.000000, 72.000000
slaw_to_v4: error -210006

===== test 12 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: 72.000000, 72.000000, 72.000000, 72.000000

===== test 13 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 14 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 15 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 16 =====

slaw_to_v2: 3.140000, 3.140000
slaw_to_v3: 3.140000, 3.140000, 0.000000
slaw_to_v4: error -210006

===== test 17 =====

slaw_to_v2: error -210006
slaw_to_v3: 3.140000, 3.140000, 3.140000
slaw_to_v4: error -210006

===== test 18 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: 3.140000, 3.140000, 3.140000, 3.140000

===== test 19 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 20 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 21 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 22 =====

slaw_to_v2: 3.141590, 3.141590
slaw_to_v3: 3.141590, 3.141590, 0.000000
slaw_to_v4: error -210006

===== test 23 =====

slaw_to_v2: error -210006
slaw_to_v3: 3.141590, 3.141590, 3.141590
slaw_to_v4: error -210006

===== test 24 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: 3.141590, 3.141590, 3.141590, 3.141590

===== test 25 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 26 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: 8.000000, 9.000000, 10.000000, 11.000000

===== test 27 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: error -210006

===== test 28 =====

slaw_to_v2: -1.000000, 0.000100
slaw_to_v3: -1.000000, 0.000100, 0.000000
slaw_to_v4: error -210006

===== test 29 =====

slaw_to_v2: error -210003
slaw_to_v3: error -210003
slaw_to_v4: error -210003

===== test 30 =====

slaw_to_v2: error -210003
slaw_to_v3: error -210003
slaw_to_v4: error -210003

===== test 31 =====

slaw_to_v2: 1.000000, 2.000000
slaw_to_v3: 1.000000, 2.000000, 0.000000
slaw_to_v4: error -210006

===== test 32 =====

slaw_to_v2: error -210006
slaw_to_v3: 3.000000, 4.000000, 5.000000
slaw_to_v4: error -210006

===== test 33 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: 6.000000, 7.000000, 8.000000, 9.000000

===== test 34 =====

slaw_to_v2: error -210003
slaw_to_v3: error -210003
slaw_to_v4: error -210003

===== test 35 =====

slaw_to_v2: error -210006
slaw_to_v3: error -210006
slaw_to_v4: 1.000000, 2.000000, 3.000000, 4.000000
